### DISCUSSION TITLE
Custom Inputs optionally used with react-hook-forms

### DISCUSSION THREAD

**Discussion by jukkahuuskonen** [react-hook-form/react-hook-form#7266] [RFC]
Hi,
In my project I have several custom inputs: Selects, CheckboxTree etc. I have written them so that you can actually use them with react-hook-forms or without it, since I rather have only one component to use whether it is inside a react-hook-form or not. This causes some problems with using the hook-form hooks, since eg. **useWatch** and **useController** require the control object or they will crash. And hooks aren't supposed to be wrapped in **try...catch** statements.

My current solution is to actually write 2 Components, one to use wihtout hook-form and the other one to with hook-forms and then have a wrapper component that checks for **control**-object and loads correct component accordingly.

I'm just thinking would it be possible to have those hooks just return some kind of error-object if the **control** is missing instead of crashing the app. That way I could simply check for those inside the component and there would be no need for 2 separate components and if I make changes to eg. their layout, I wouldn't need to do that to 2 components.

**Comment by JamesWong1999** [comment]
can you share a simple codesandbox?

**Comment by jukkahuuskonen** [comment]
Here is a super [Simplified CodeSandbox Example](https://codesandbox.io/s/falling-cache-ew93e?file=/src/App.js:685-819).

This version would not actually be a problem to implement by just using **<Controller />** conditionally depending whether **control** is passed to the component, but it demonstrates the same problem that actually exists for more complicated components.
If there is no **control** passed to **MyCheckbox**, then the **useController** (and other hooks depending on **control**) will crash the component.
Since it is not advisable (not sure if it is possible at all) to wrap hooks in **try...catch**-statements, I'll basically need to create 2 sets of components, one for hook forms and one for without it. Depending on how complicated the component is, this may mean lot of things to fix in 2 places when updating components. I would prefer having these hooks (useController, useWatch, ...) that require **control** to return an error object or just empty object instead of crashing the whole app. Then implementing these kind of components would be easier.

In the pic below the **MyCheckbox**-component that has been commented out will crash the app, since it doesn't have **control**-prop (or **name**). **MyCheckbox** has been written, so that if the **useController**-hook wouldn't crash without **control** and returned just an empty object, it would work just as a normal **MuiCheckbox**.
![image](https://user-images.githubusercontent.com/42405685/145978686-44c822a2-5fc6-4583-ad12-db11f151e5a6.png)
